[libpas] Land PGM stack trace collection

zakariaridouh opened the pull request:
#### 21f7771c163dfa0727792de83714d754b5c97122
<pre>
[libpas] Land PGM stack trace collection
<a href="https://bugs.webkit.org/show_bug.cgi?id=287049">https://bugs.webkit.org/show_bug.cgi?id=287049</a>
<a href="https://rdar.apple.com/106166552">rdar://106166552</a>

Reviewed by Yusuke Suzuki.

Reland stack trace collection for PGM allocations and deallocations
after A/B performance testing came back neutral.

Below is the original commit message for additional context:

This adds collection and storage of &lt;=31 frames of backtrace upon PGM
allocation and deallocation. A followup patch will pack these frame
pointers more aggressivly.

Additionally, bump `pas_crash_report_version` to version 3 so
`reportcrashd` is aware of struct changes.

Add a variety of tests to validate this new functionality.

* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
(backtrace):
(pas_probabilistic_guard_malloc_allocate):
(pas_probabilistic_guard_malloc_deallocate):
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.h:
* Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
(pas_update_report_crash_fields):
(pas_report_crash_extract_pgm_failure):
* Source/bmalloc/libpas/src/libpas/pas_report_crash_pgm_report.h:
* Source/bmalloc/libpas/src/libpas/pas_segmented_vector.h:
* Source/bmalloc/libpas/src/libpas/pas_utils.h:
(pas_non_empty_ranges_overlap):
(pas_ranges_overlap):
* Source/bmalloc/libpas/src/test/EnumerationTests.cpp:
(std::testPGMEnumerationBasic):
(std::testPGMEnumerationAddAndFree):
* Source/bmalloc/libpas/src/test/PGMTests.cpp:
(std::testPGMMultipleAlloc):
(std::testPGMMetadataVectorManagementRehash):
(std::testPGMBmallocAllocationBacktrace):
(std::testPGMAllocMetadataOnly):
(addPGMTests):

Canonical link: <a href="https://commits.webkit.org/294039@main">https://commits.webkit.org/294039@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b777bd3dccea6c718c11f2c783c59245b72e4d73

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100657 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20309 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10608 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105794 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51245 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20617 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28783 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76648 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33683 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103664 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15826 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90916 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57003 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15635 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8923 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50621 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/93321 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85547 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8998 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108149 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/99265 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27775 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/20394 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85601 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28138 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87117 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85142 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21661 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29842 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7569 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21763 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27710 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/122891 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27521 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34265 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30839 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29079 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
```diff
@@ -181,15 +193,16 @@ pas_allocation_result pas_probabilistic_guard_malloc_allocate(pas_large_heap* la
-    value->size_of_data_pages          = mem_to_alloc - (lower_guard_size + upper_guard_size);
-    value->start_of_data_pages         = result.begin + lower_guard_size;
+    value->alloc_backtrace.frame_size  = backtrace(value->alloc_backtrace.backtrace_buffer, PGM_BACKTRACE_MAX_FRAMES);
+    value->mem_to_waste                 = mem_to_waste;
```
It is not well formatted.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.h:
libpas does not use this form. Define struct backtrace_metadata and use it as " struct backtrace_metadata" Also it should be prefixed with pas_

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
```diff
@@ -58,12 +58,14 @@ static PAS_ALWAYS_INLINE bool PAS_WARN_UNUSED_RETURN pas_fault_address_is_in_bou
 }
 
-static PAS_ALWAYS_INLINE kern_return_t PAS_WARN_UNUSED_RETURN pas_update_report_crash_fields(pas_report_crash_pgm_report* report, const char* error_type, const char* confidence, vm_address_t fault_address, size_t allocation_size)
+static PAS_ALWAYS_INLINE kern_return_t PAS_WARN_UNUSED_RETURN pas_update_report_crash_fields(pas_report_crash_pgm_report* report, const char* error_type, const char* confidence, vm_address_t fault_address, size_t allocation_size, backtrace_metadata_t alloc_backtrace, backtrace_metadata_t dealloc_backtrace)
```
it should get "struct backtrace_metadata*"

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_report_crash_pgm_report.h:
```diff
@@ -54,6 +56,8 @@ typedef struct {
     size_t allocation_size;
+    backtrace_metadata_t alloc_backtrace;
+    backtrace_metadata_t dealloc_backtrace;
 } pas_report_crash_pgm_report;
```
We should not have this typedef form.

Constellation commented on Source/bmalloc/libpas/src/test/PGMTests.cpp:
```diff
@@ -419,7 +427,7 @@ void testPGMMetadataDoubleFreeBehavior()
 void testPGMMetadataVectorManagementRehash()
 {
-    pas_probabilistic_guard_malloc_initialize_pgm_as_enabled(1);
+    pas_probabilistic_guard_malloc_initialize_pgm_as_enabled();
```
This is wrong. Can you ensure this test gets build and pass?

zakariaridouh replied:
Leftover from reapplying the previous patch. Fixed.

Constellation commented on Source/bmalloc/libpas/src/test/PGMTests.cpp:
```diff
@@ -474,6 +482,107 @@ void testPGMMetadataVectorManagementRehash()
     pas_heap_lock_unlock();
 }
 
+void testPGMBmallocAllocationBacktrace() {
```
Bot formatted correctly.

Constellation commented on Source/bmalloc/libpas/src/test/PGMTests.cpp:
```diff
@@ -474,6 +482,107 @@ void testPGMMetadataVectorManagementRehash()
+    }
+}
+
+void testPGMAllocMetadataOnly() {
```
Ditto

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
```diff
@@ -58,12 +58,14 @@ static PAS_ALWAYS_INLINE bool PAS_WARN_UNUSED_RETURN pas_fault_address_is_in_bou
 }
 
-static PAS_ALWAYS_INLINE kern_return_t PAS_WARN_UNUSED_RETURN pas_update_report_crash_fields(pas_report_crash_pgm_report* report, const char* error_type, const char* confidence, vm_address_t fault_address, size_t allocation_size)
+static PAS_ALWAYS_INLINE kern_return_t PAS_WARN_UNUSED_RETURN pas_update_report_crash_fields(pas_report_crash_pgm_report* report, const char* error_type, const char* confidence, vm_address_t fault_address, size_t allocation_size, struct pas_backtrace_metadata* alloc_backtrace, struct pas_backtrace_metadata* dealloc_backtrace)
```
Remove `struct` from `pas_backtrace_metadata`.

zakariaridouh committed:
[libpas] Land PGM stack trace collection
https://bugs.webkit.org/show_bug.cgi?id=287049
rdar://106166552

Reviewed by Yusuke Suzuki.

Reland stack trace collection for PGM allocations and deallocations
after A/B performance testing came back neutral.

Below is the original commit message for additional context:

This adds collection and storage of <=31 frames of backtrace upon PGM
allocation and deallocation. A followup patch will pack these frame
pointers more aggressivly.

Additionally, bump `pas_crash_report_version` to version 3 so
`reportcrashd` is aware of struct changes.

Add a variety of tests to validate this new functionality.

* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
(backtrace):
(pas_probabilistic_guard_malloc_allocate):
(pas_probabilistic_guard_malloc_deallocate):
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.h:
* Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
(pas_update_report_crash_fields):
(pas_report_crash_extract_pgm_failure):
* Source/bmalloc/libpas/src/libpas/pas_report_crash_pgm_report.h:
* Source/bmalloc/libpas/src/libpas/pas_segmented_vector.h:
* Source/bmalloc/libpas/src/libpas/pas_utils.h:
(pas_non_empty_ranges_overlap):
(pas_ranges_overlap):
* Source/bmalloc/libpas/src/test/EnumerationTests.cpp:
(std::testPGMEnumerationBasic):
(std::testPGMEnumerationAddAndFree):
* Source/bmalloc/libpas/src/test/PGMTests.cpp:
(std::testPGMMultipleAlloc):
(std::testPGMMetadataVectorManagementRehash):
(std::testPGMBmallocAllocationBacktrace):
(std::testPGMAllocMetadataOnly):
(addPGMTests):

Canonical link: https://commits.webkit.org/294039@main
